test/generated_cases_test.exs

```diff
@@ -37,6 +37,10 @@ defmodule PublicSuffix.TestCaseGenerator do
       input: input,
       registrable_domain_output: registrable_domain_output,
       public_suffix_output: public_suffix_output(registrable_domain_output, input),
+      # prevailing_rule/2 should always return the same as public_suffix/2 except
+      # when the domain doesn't match an explicit rule, so the following should be fine,
```

Comment:
> prevailing_rule/2 should always return the same as public_suffix/2 except
> when the domain doesn't match an explicit rule

I don't think this is true: there are rules like [`*.bd`](https://github.com/seomoz/publicsuffix-elixir/blob/887d7e0c1650c46d8ce29564b9e3f0c36b5f12f4/data/public_suffix_list.dat#L297) and [`!www.ck`](https://github.com/seomoz/publicsuffix-elixir/blob/887d7e0c1650c46d8ce29564b9e3f0c36b5f12f4/data/public_suffix_list.dat#L583, which are clearly different than what the matching public suffix will be.

Actually, your writing this comment made me realize that our implementation of `prevailing_rule/2` is buggy!  Specifically, the `find_prevailing_rule/2` helper function does _not_ return a rule (except in the specific case when no explicit rule matches, and then it returns `["*"]`).  It returns the domain parts that match the prevailing rule.  For example, the `@wild_card_rules[["*" | suffix]] in allowed_rule_types -> domain_labels` clearly returns the labels from the input domain, not the wild card rule itself.  Similarly, `find_prevailing_exception_rule` does not return the rule, but instead returns the part of the domain that the algorithm says to count as the public suffix according to the rule.

To properly support `prevailing_rule/2` I think we need to actually have `find_prevailing_rule` return the matching rule, but to do that I think the return type needs to be richer so the caller can distinguish between an exception rule and a normal rule (which affects how the rule is used).  Maybe we can refactor things so that a tuple like `{:normal, ["*", "bd"]}` or `{:exception, ["www", "ck"]}` is returned by the private `find_prevailing_rule/2` function?  I think that would support both the needs of the existing public APIs as well as this new function you are adding.

Thoughts?